README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# svgpatternusgs <img src="man/figures/logo.svg" align="right" height=230/>

<!-- badges: start -->

![](http://img.shields.io/badge/cool-useless-green.svg)
![](http://img.shields.io/badge/mini-verse-blue.svg)
<!-- badges: end -->

`svgpatternusgs` provides SVG patterns from the [United States
Geological Survey (USGS)](usgs.gov).

For online documentation, see the [svgpatternusgs pkgdown
website](https://coolbutuseless.github.io/package/svgpatternusgs/index.html).

The USGS provides a large array of reference styles for geologic
linework and map symbology.

  - [USGS website for Geological Map
    Symbols](https://ngmdb.usgs.gov/fgdc_gds/geolsymstd/download.php)
  - The raw data for this package was sourced from [davenquinn’s github
    version](https://github.com/davenquinn/geologic-patterns)

## An example of the patterns provided

This is a screenshot of the original documentation which comes with
these patterns from the USGS

<img src="man/figures/USGS-sample.png" />

This list of all codes which have a pattern in this package:

``` r
svgpatternusgs::all_usgs_codes
#>   [1] 101 102 103 104 105 106 114 116 117 118 119 120 121 122 123 124 132 134
#>  [19] 135 136 137 201 202 204 206 207 214 215 216 217 218 219 226 228 229 230
#>  [37] 231 232 233 301 302 303 304 305 306 313 314 315 316 317 318 319 327 328
#>  [55] 330 331 401 402 403 405 406 411 412 416 417 418 419 420 423 424 427 428
#>  [73] 429 430 431 432 433 434 435 436 521 522 523 524 591 592 593 594 595 601
#>  [91] 602 603 605 606 607 608 609 610 611 612 613 614 616 617 618 619 620 621
#> [109] 622 623 624 625 626 627 628 629 630 631 632 633 634 635 636 637 638 639
#> [127] 640 641 642 643 644 645 646 647 648 649 650 651 652 653 654 655 656 657
#> [145] 658 659 660 661 662 663 664 665 666 667 668 669 670 671 672 673 674 675
#> [163] 676 677 678 679 680 681 682 683 684 685 686 701 702 703 704 705 706 707
#> [181] 708 709 710 711 712 713 714 715 716 717 719 720 721 722 723 724 725 726
#> [199] 727 728 729 730 731 732 733
```

## What’s in the box?

  - Functions for accessing the SVG for all the provided USGS codes
  - SVG pattern are supplied as `minisvg::SVGPattern` objects.
  - Functions for encoding/decoding a pattern specification from a hex
    colour.

## To Do

The original SVG sources for this were OK but not perfect. I’ve patched
some issues, but there are still quite a few occasions where gaps and
tears appear in patterns.

These errors don’t appear to be systematic, so I think it will require
some labour-intensive manual corrections to individual patterns.

## Installation

You can install from
[github](https://github.com/coolbutuseless/svgpatternusgs) with:

``` r
# install.packages("devtools")
devtools::install_github("coolbutuseless/lofi")           # Colour encoding
devtools::install_github("coolbutuseless/minisvg")        # SVG support
devtools::install_github("coolbutuseless/svgpatternusgs") # This package
```

## Create and show a pattern

``` r
pattern <- create_usgs_pattern(usgs_code = 605, spacing = 50)

# pattern$show()
pattern$save_full_svg("man/figures/usgs-605.svg", width=400, height = 100)
```

<img src="man/figures/usgs-605.svg" />

### Multiple patterns

``` r
pattern_list <- svgpatternusgs::all_usgs_codes[41:72] %>%
  purrr::map(create_usgs_pattern, spacing = 200) 

patterns <- minisvg::SVGPatternList_to_svg(pattern_list, width=200, height=100, ncol = 4)

patterns$save("man/figures/usgs-sample.svg")
```

<img src="man/figures/usgs-sample.svg" />

## Using the pattern in a `minisvg` SVG document

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Building an SVG logo with an animated stripe
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
library(minisvg)
doc <- minisvg::svg_doc(width = 400, height = 200)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create a simple pattern using the `svgpatternsimple` package
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
geopattern1 <- svgpatternusgs::create_usgs_pattern(usgs_code = 601, spacing = 200)
geopattern2 <- svgpatternusgs::create_usgs_pattern(usgs_code = 655, spacing = 200)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Add the pattern to the documents list of definitions
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
doc$defs(geopattern1, geopattern2)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Create shapes filled with the patterns and add to doc
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rect <- stag$rect(x='10%', y='10%', width='50%', height='50%', stroke = 'black', 
                  fill=geopattern1)

circle <- stag$circle(cx = "75%", cy="60%", r="45%", stroke = 'black', 
                      fill=geopattern2)

doc$append(rect, circle)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# output
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# doc$show()
doc$save("man/figures/README-example-rect.svg")
```

<pre><details><summary style='color: #4169E1;'> Show/hide SVG text </summary>&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;svg width="400" height="200" viewBox="0 0 400 200" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"&gt;
  &lt;defs&gt;
    &lt;pattern id="usgs-601-ffffff-1-0" patternUnits="userSpaceOnUse" patternTransform="rotate(0 0 0)" x="0px" y="0px" width="200" height="200" viewBox="1.888 -55.518 54.125 54.125"&gt;
      &lt;g&gt;
        &lt;rect x="1.888" y="-55.518" style="fill:#ffffff; fill-opacity:1; stroke:none;" width="54.125" height="54.125" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="27.105" y1="-29.17" x2="27.105" y2="-29.17" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="2.72" y1="-43.255" x2="2.72" y2="-43.255" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="5.25" y1="-41.825" x2="5.25" y2="-41.825" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="7.945" y1="-43.585" x2="7.945" y2="-43.585" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="7.945" y1="-40.505" x2="7.945" y2="-40.505" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="11.3" y1="-42.595" x2="11.3" y2="-42.595" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="14.27" y1="-43.64" x2="14.27" y2="-43.64" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="11.355" y1="-39.075" x2="11.355" y2="-39.075" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="14.733" y1="-37.067" x2="14.733" y2="-37.067" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="14.567" y1="-40.233" x2="14.567" y2="-40.233" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="17.13" y1="-42.375" x2="17.13" y2="-42.375" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="20.212" y1="-44.4" x2="20.212" y2="-44.4" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="6.075" y1="-38.415" x2="6.075" y2="-38.415" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="3.05" y1="-38.91" x2="3.05" y2="-38.91" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="4.535" y1="-35.94" x2="4.535" y2="-35.94" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="7.67" y1="-35.28" x2="7.67" y2="-35.28" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="10.915" y1="-35.39" x2="10.915" y2="-35.39" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="18.12" y1="-37.535" x2="18.12" y2="-37.535" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="16.733" y1="-34.567" x2="16.733" y2="-34.567" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="13.995" y1="-33.245" x2="13.995" y2="-33.245" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="19.66" y1="-34.455" x2="19.66" y2="-34.455" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="21.53" y1="-37.315" x2="21.53" y2="-37.315" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="20.485" y1="-41.66" x2="20.485" y2="-41.66" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="11.275" y1="-46.088" x2="11.275" y2="-46.088" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="24.83" y1="-43.2" x2="24.83" y2="-43.2" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="23.455" y1="-40.45" x2="23.455" y2="-40.45" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="24.5" y1="-37.315" x2="24.5" y2="-37.315" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="22.96" y1="-34.345" x2="22.96" y2="-34.345" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="26.865" y1="-40.065" x2="26.865" y2="-40.065" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="28.845" y1="-42.65" x2="28.845" y2="-42.65" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="28.02" y1="-37.37" x2="28.02" y2="-37.37" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="26.205" y1="-34.18" x2="26.205" y2="-34.18" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="30.55" y1="-39.13" x2="30.55" y2="-39.13" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="2.94" y1="-33.135" x2="2.94" y2="-33.135" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="6.13" y1="-32.64" x2="6.13" y2="-32.64" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="9.375" y1="-32.42" x2="9.375" y2="-32.42" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="32.42" y1="-42.045" x2="32.42" y2="-42.045" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="32.15" y1="-36.9" x2="32.15" y2="-36.9" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="34.9" y1="-38.4" x2="34.9" y2="-38.4" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="33.41" y1="-33.685" x2="33.41" y2="-33.685" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="36.765" y1="-34.62" x2="36.765" y2="-34.62" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="38.195" y1="-37.975" x2="38.195" y2="-37.975" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="38.085" y1="-42.485" x2="38.085" y2="-42.485" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="41.165" y1="-39.625" x2="41.165" y2="-39.625" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="43.145" y1="-42.65" x2="43.145" y2="-42.65" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="45.235" y1="-39.515" x2="45.235" y2="-39.515" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="42.43" y1="-36.215" x2="42.43" y2="-36.215" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="40.065" y1="-33.465" x2="40.065" y2="-33.465" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="46.72" y1="-42.375" x2="46.72" y2="-42.375" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="50.65" y1="-46.4" x2="50.65" y2="-46.4" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="53.9" y1="-49.15" x2="53.9" y2="-49.15" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="54.4" y1="-45.15" x2="54.4" y2="-45.15" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="51.525" y1="-50.65" x2="51.525" y2="-50.65" /&gt;
        &lt;line style="fill:none;stroke:#000000;stroke-width:0.7;stroke-linecap:round;" x1="48.4" y1="-48.15" x2="48.4" y2="-48.15" /&gt;

NEWS.md

# svgpatternusgs 0.2.0  2020-02-01

* Remove old `{devoutsvg-0.1.0}` compatibility. 
  
# svgpatternusgs 0.1.0

* Initial release


data-raw/90-prepare-svg-files.R

rm(list=ls())

library(xml2)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# filenames for the raw SVG patterns
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
svg_files <- list.files("data-raw/geologic-patterns/assets/svg/", full.names = TRUE)

svg_files <- svg_files[grepl("-K", svg_files) | nchar(basename(svg_files))==7]
svg_file <- svg_files[1]

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# 01 - Whitespace tidy
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
outdir01 <- "data-raw/01-svg-whitespace"
dir.create(outdir01, showWarnings = FALSE)

for (svg_file in svg_files) {
  svg_text <- readLines(svg_file)
  svg_text <- stringr::str_trim(svg_text)
  svg_text <- svg_text[svg_text != '']
  svg_text <- gsub("\\s+", " ", svg_text)

  outfile <- file.path(outdir01, basename(svg_file))
  svg_obj <- xml2::read_xml(paste(svg_text, collapse="\n"))
  xml2::write_xml(svg_obj, outfile)
}




#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# 02 - Convert to raw <pattern></pattern> rather than <svg></svg>
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
svg_files <- list.files(outdir01, full.names = TRUE)
svg_file  <- svg_files[4]
outdir02  <- "data-raw/02-svg-pattern-raw"
dir.create(outdir02, showWarnings = FALSE)


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Build a generic rectangle to use as the background for every pattern
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# bg_rect <- xml2::read_xml("<rect width='100%' height='100%' style='fill:#ffffff; fill-opacity:1.00; stroke:none;'/>")


for (svg_file in svg_files) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Read the SVG object, take note of its attributes (to apply to pattern later)
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  svg_obj   <- read_xml(svg_file)
  svg_attrs <- xml_attrs(svg_obj)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Every pattern needs a viewbox so we can scale it later for display
  # by maniupulating the 'width' and 'height' attributes.
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  if (!'viewBox' %in% names(svg_attrs)) {
    svg_attrs['viewBox'] <- as.character(glue::glue("0 0 {svg_attrs['width']} {svg_attrs['height']}"))
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # throw away the <svg> wrapper and just keep the inner
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  g_obj     <- xml_children(svg_obj)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Create an empty pattern with the correct attributes
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  id <- gsub("-K", "", tools::file_path_sans_ext(basename(svg_file)))
  id <- paste0("usgs-", id)

  # patternTransform='rotate({angle} 0 0)'

  pattern_attrs <- c()
  pattern_attrs['id'] <- id
  pattern_attrs['patternUnits']     <- 'userSpaceOnUse'
  pattern_attrs['patternTransform'] <- 'rotate(0 0 0)'
  pattern_attrs <- c(pattern_attrs, svg_attrs[intersect(c('x', 'y', 'width', 'height', 'viewBox'), names(svg_attrs))])
  pattern_node <- xml2::read_xml("<pattern></pattern>")
  xml_set_attrs(pattern_node, pattern_attrs)


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # some files are missing a BG rect
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  lines <- readLines(svg_file)
  rects <- grep("rect.*?fill:none", lines)
  if (length(rects) == 0) {
    # This file is missing a bg rect
    message("Adding bg_rect for: ", svg_file)

    viewbox <- pattern_attrs[['viewBox']]
    viewbox <- strsplit(viewbox, "\\s+")[[1]]  # x, y, w, h
    bg_rect <- glue::glue('<rect x="{viewbox[1]}" y="{viewbox[2]}" style="fill:none; stroke:none;" width="{viewbox[3]}" height="{viewbox[4]}"/>')
    bg_rect <- xml2::read_xml(bg_rect)
    xml_add_child(pattern_node, bg_rect)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Copy across all the svg nodes into this pattern
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  for (x in g_obj) {
    xml_add_child(pattern_node, x)
  }

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Strip the default 'namespace' attribute and output to file.
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  outfile <- file.path(outdir02, basename(svg_file))
  outfile <- gsub("-K", "", outfile)
  write_xml(xml_ns_strip(pattern_node),
            outfile,
            options = c('format', 'no_declaration'))

}




#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# isolated rect corrections
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
filename <- "data-raw/02-svg-pattern-raw/104.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<path fill="none" d="M0 0h33.5v33.5H0z"/>',
                 '', svg_text)
writeLines(svg_text, filename)

filename <- "data-raw/02-svg-pattern-raw/135.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<path fill="none" d="M0 0h53.18v53.6H0z"/>',
                 '', svg_text)
writeLines(svg_text, filename)

filename <- "data-raw/02-svg-pattern-raw/202.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<path d="M0 0h73.86v73.5H0z"/>',
                 '', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/215.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="0" y="-74.5" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;" width="74.499" height="74.499"/>',
                 '<rect x="0" y="-74.5" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;stroke:none;" width="74.499" height="74.499"/>', svg_text)
writeLines(svg_text, filename)



filename <- "data-raw/02-svg-pattern-raw/218.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="1" y="-73.66" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;" width="72" height="72"/>',
                 '<rect x="1" y="-73.66" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;stroke:none;" width="72" height="72"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/219.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="1" y="-79.444" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;" width="77.784" height="77.785"/>',
                 '<rect x="1" y="-79.444" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;stroke:none;" width="77.784" height="77.785"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/229.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="1" y="-75.088" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;" width="73.428" height="73.428"/>',
                 '<rect x="1" y="-75.088" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;stroke:none;" width="73.428" height="73.428"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/230.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="0" y="-80.646" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;" width="80.646" height="80.646"/>',
                 '<rect x="0" y="-80.646" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;stroke:none;" width="80.646" height="80.646"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/231.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="0" y="-90.305" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;" width="90.306" height="90.305"/>',
                 '<rect x="0" y="-90.305" style="fill-rule:evenodd;clip-rule:evenodd;fill:none;stroke:none;" width="90.306" height="90.305"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/402.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="0" y="0" style="fill:none;" width="71.339" height="71.301"/>',
                 '<rect x="0" y="0" style="fill:none;stroke:none;" width="71.339" height="71.301"/>', svg_text)
svg_text <- gsub('<path fill="none" d="M0 0h71.34v71.3H0z"/>',
                 '', svg_text)
writeLines(svg_text, filename)



filename <- "data-raw/02-svg-pattern-raw/427.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="1" y="-8.924" style="fill:none;" width="7.924" height="7.924"/>',
                 '<rect x="1" y="-8.924" style="fill:none;stroke:none;" width="7.924" height="7.924"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/602.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="2.077" y="-55.938" style="fill:none;" width="54.125" height="54.125"/>',
                 '<rect x="2.077" y="-55.938" style="fill:none; stroke:none;" width="54.125" height="54.125"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/612.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect y="-54.125" style="fill:none;" width="54.125" height="54.125"/>',
                 '<rect y="-54.125" style="fill:none;stroke:none;" width="54.125" height="54.125"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/630.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="0.043" y="-54.125" style="fill:none;" width="54.125" height="54.125"/>',
                 '<rect x="0.043" y="-54.125" style="fill:none;stroke:none;" width="54.125" height="54.125"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/658.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect y="-54.125" width="54.125" height="54.125"/>',
                 '<rect y="-54.125" width="54.125" height="54.125" style="fill:none;stroke:none;"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/659.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="0.107" y="-54.125" width="54.125" height="54.125"/>',
                 '<rect x="0.107" y="-54.125" width="54.125" height="54.125" style="fill:none;stroke:none;"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/668.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect y="-54.125" style="fill:none;" width="54.125" height="54.125"/>',
                 '<rect y="-54.125" style="fill:none;stroke:none;" width="54.125" height="54.125"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/676.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect y="-54.125" style="fill:none;" width="54.125" height="54.125"/>',
                 '<rect y="-54.125" style="fill:none;stroke:none;" width="54.125" height="54.125"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/680.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect x="0" y="-48.712" style="fill:none;" width="48.712" height="48.712"/>',
                 '<rect x="0" y="-48.712" style="fill:none;stroke:none;" width="48.712" height="48.712"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/681.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect y="-54.125" style="fill:none;" width="54.125" height="54.125"/>',
                 '<rect y="-54.125" style="fill:none;stroke:none;" width="54.125" height="54.125"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/684.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect y="-41.881" style="fill:none;" width="41.881" height="41.881"/>',
                 '<rect y="-41.881" style="fill:none;stroke:none;" width="41.881" height="41.881"/>', svg_text)
writeLines(svg_text, filename)


filename <- "data-raw/02-svg-pattern-raw/685.svg"
svg_text <- readLines(filename)
svg_text <- gsub('<rect y="-54.125" style="fill:none;" width="54.125" height="54.125"/>',
                 '<rect y="-54.125" style="fill:none;stroke:none;" width="54.125" height="54.125"/>', svg_text)
writeLines(svg_text, filename)






if (FALSE) {
  library(svgpatternusgs)
  moar <- list.files(outdir02, full.names = TRUE, pattern = "*svg")
  pattern <- svgpatternusgs:::read_pattern(moar[6])


  pattern_files <- moar[1:25]
  patterns <- purrr::map(pattern_files, svgpatternusgs:::read_pattern)
}

















